### PR TITLE
serial 1wire passthrough for blheli 14.0.0.3.2+

### PR DESCRIPTION
**Pull request by nathantsoi**
@4712 put lots of hard work into this and I'd love to see it merged in soon. It looks like the next version of BLHeliSuite will supports the MSP interface (beta below) and I've tested it on my NAZE and CC3D. More testing/other hardware would be great.

All Binaries are here (latest master as of 9/30/2015 + this pull request):

https://www.dropbox.com/sh/4el0i13obglt5ot/AABStBuLr4hBXL414EfqCRSWa?dl=0

Per @4712's instructions below, download the BLHeliSuite.exe 1.4.1.0.3.1 and replace it with the existing BLHeliSuite.exe in the BLHeliSuite 1.4.1.0.3 folder. 
http://www.mediafire.com/download/5yd45y0e2st4mry/BLHeliSuite141031.7z

Walkthrough: http://youtu.be/rf_HTdQfQhE

supersedes: https://github.com/cleanflight/cleanflight/pull/1159

completes: https://github.com/cleanflight/cleanflight/issues/216

### CODE REVIEW

**Comment by 4712** [comment]
@nathantsoi 
please remove line 2138 to 39 in serial_cli.c
    UNUSED(cmdline);
    StopPwmAllMotors();

(StopPwmAllMotors(); doesn't do anything for me)


and please remove line 1758 to 60 in serial_msp.c

 // Client should check active Motors before preinitialize the Passthrough
 // with MSP_MOTOR and check each value >0 and later call only active channels
 // rem: atm not allowed channel mapping other than 0..x ascending

(comment is invalid)

File: docs/1wire.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+This is the option you need to select for the bootloader:
+
+![Flashing BlHeli Bootloader](assets/images/blheli-bootloader.png)
+
+Currently supported on the SPRACINGF3, STM32F3DISCOVERY, NAZE32 (including clones such as the FLIP32) and CC3D.
+
+## Wiring
+
+  - For the NAZE, no external wiring is necessary. Simply plugin the board via USB cable.
```
Comment (Echelon9) [inline docs/1wire.md]: s/plugin/plug in/

File: src/main/io/serial_1wire.c [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  gpio_config_t cfg;
+  cfg.pin = pin;
+  cfg.mode = mode;
+  cfg.speed = Speed_10MHz;
+  gpioInit(gpio, &cfg);
+}
+
+void usb1WireInitialize()
+{
+   for (volatile uint8_t i = 0; i<ESC_COUNT ; i++){
```
Comment (Echelon9) [inline src/main/io/serial_1wire.c]: For consistent spacing, should be s/i<ESC_COUNT ;/i < ESC_COUNT;/

File: src/main/io/serial_msp.c [hunk 3]
```diff
@@ -1721,6 +1725,47 @@ static bool processInCommand(void)
...
+    case MSP_SET_1WIRE:
+        // get channel number
+        i = read8();
+        // we do not give any data back, assume channel number is transmitted OK
+        if (i==0xFF) {
+            // 0xFF -> preinitialize the Passthrough
+            // switch all motor lines HI
+            usb1WireInitialize();
+            // and come back rigth afterwards
+            // rem: App: Wait at least appx. 500 ms for BLHeli to jump into bootloader mode before try to connect any ESC
```
Comment (Echelon9) [inline src/main/io/serial_msp.c]: Preferable to wrap comment here onto a new line after c.80 characters, so easier to read.

**Comment by Fheilmann** [comment]
Any plans on extending this functionality for the SimonK bootloader as well? A lot of BLHeli flashed ESCs shipped (DYS being a prominent one) come with BLHeli FW but a SimonK bootloader using the STK500v2 protocol I believe. This would remove the need to remove the heatshrink and solder an ISP interface to those ESCs to make them compatible with this method

File: src/main/io/serial_msp.c (hunk 3 above)
Comment (nathantsoi) [inline src/main/io/serial_msp.c]: Awesome! Thx for the code review @4712 and @Echelon9, I will update tonight. @Echelon9 is a force-pushed update to this commit or another commit preferred?

Comment (Echelon9) [inline src/main/io/serial_msp.c]: I don't have a preference for one approach over the other.

**Comment by nathantsoi** [comment]
ok, this should be ready to go @hydra. @4712 can confirm.

**Comment by 4712** [comment]
@nathantsoi: confirm, looks good to me.

**Comment by Echelon9** [comment]
LGTM

**Comment by 4712** [comment]
I can check any motor is running but how can I Implement reading ARMING_FLAG(ARMED) with any MSP command into BLHeliSuite?

**Comment by thenickdude** [comment]
@4712 the MSP_STATUS command includes that information, the arming status is the lowest bit in the 32-bit response in that command.

**Comment by 4712** [comment]
@thenickdude: Thanks a lot!
I tried MSP_STATUS before, but got always 0 for the whole junk. But now it works...

**Comment by borisbstyle** [comment]
@nathantsoi 
Can you check why I have build errors for SPRACINGF3with this in betaflight PR you made?

**Comment by borisbstyle** [comment]
You need to reallocate this function to somewhere earlier in the code

static void mspProcessReceivedCommand() {

**Comment by 4712** [comment]
@borisbstyle:
usb1WireInitialize was positioned inside an ifdef clause 
Fix: https://github.com/4712/cleanflight/commit/4ef16b15f9fc647347161c4df2cb6dd855e58ae7

But I get an error compiling betaflight local
rx.c line 86 "<<<<<<< HEAD"
Removing this solved the problem for me

**Comment by borisbstyle** [comment]
@4712 
Thanks for the fix. rx.c is also fixed now

**Comment by hydra** [comment]
Hi everyone.

So I had a thought regarding the use of the GPIO configuration pins definitions that are a problem here.

1) it's duplicate information (bad)
2) it requires extra maintenance (bad)
3) it's more target specific code (which we don't want).

There are many solutions of course.  One being the big IO allocation cleanup, see #357.

Short term however it would be simple enough to store the uses motor pins and timer configuration references in an extended version of `pwmOutputConfiguration_t` which the onewire code could ask for e.g. `pwmGetOutputConfiguration()`.

example configuration structure.

```
typedef enum {
  PWM_PF_NONE = 0;
  PWM_PF_MOTOR (1 << 0) 
  PWM_PF_SERVO (1 << 1) 
  PWM_PF_MOTOR_MODE_BRUSHED (1 << 2)
  PWM_PF_OUTPUT_PROTOCOL_PWM (1 << 3)
  PWM_PF_OUTPUT_PROTOCOL_ONESHOT (1 << 4)
} pwmPortFlags_e;

typedef struct pwmPortConfiguration_s {
    uint8_t index;
    pwmPortFlags_e flags;
    timerHardware_t *timerHardware;
} pwmPortConfiguration_t;

typedef struct pwmOutputConfiguration_s {
    uint8_t servoCount;
    uint8_t motorCount;
    pwmPortConfiguration_t portConfigurations[MAX_PWM_OUTPUT_PORTS];
} pwmOutputConfiguration_t;
```

using that the code should be able to find the GPIO Port and Pin and no additional mapping other than that already defined in timer.c is required.

**Comment by nathantsoi** [comment]
Sweet, cleanup would be great! For now, I vote we merge this as is, since it's working and in the current structure of the code, makes sense. We can pay down the technical debt later, but there is no sense in delaying a good feature while it's re-written.

**Comment by 4712** [comment]
@hydra +1
As long, as there will be no additional lag introduced into the port handling...
ESC numbers should be mapped ascending to the active ESC ports.  
I'd like to have the possibility to ask through MSP for the info, which ESC number (1..8) is valid. I guess MSP_MOTOR_PINS should theoretically exist for this. :)

**Comment by Arakon** [comment]
This appears not to work on a CC3D (anymore?). I flashed the latest betaflight (as well as a selfcompiled version with these PRs). CC3D times out trying to connect or directly gives an error. I tried using main port, flex port (MSP is active on main), using the dual USB + main port combination. They connect to CF fine, but then can not connect to the actual ESC.
The same ESC, same version of CF on a flip32+ will immediately connect with no issue whatsoever.

**Comment by nathantsoi** [comment]
You actually only need 1 USB connection now @Arakon, a USB/UART converter must be plugged into the flex port, as shown at 1:10 in this video: https://youtu.be/rf_HTdQfQhE

Also, make sure you have MSP enabled on the flex port.

Let me know how it goes

**Comment by Arakon** [comment]
Ah! I tried on the main port. I just tried with flex port and while not always recognizing the ESCs first try, it works.
Your tutorial contradicts itself then, though..

>For the CC3D, connect a USB to UART adapter to the **main port**. If you need one, I prefer the >CP2102 as it is cheap and the driver is readily available.

If possible, could the main port be used? Since it's also required for flashing, it'd be much easier to keep MSP on the main port + USB port rather than having to switch.

**Comment by 4712** [comment]
BLHeliSuite 14.1.0.0 with support for Cleanflight now posted...
https://blhelisuite.wordpress.com/

**Comment by nathantsoi** [comment]
Awesome, @4712!

Thx for the heads up @Arakon, I've updated the docs. Did you try the new blheli suite @4712 just posted? CC3D works fine for me on my miniquad with HK 12a blueseries escs.

**Comment by Arakon** [comment]
Same thing, only detects ESC #1.

**Comment by Arakon** [comment]
I just tried the hex you linked in the first post here, with exactly the same results.

**Comment by simonvc** [comment]
Is this supposed to work with the SPRF3? I've had no luck connecting to the ESCs, its obviously doing something the ESCs go beep beep beep and blhelisuite just says Connecting (power cycle) until you hit cancel and then it stops beeping.

**Comment by 4712** [comment]
Sounds like you have the Simonk bootloder installed....

**Comment by nathantsoi** [comment]
Strange @Arakon, feel free to email me a video or something and we can troubleshoot.

**Comment by Arakon** [comment]
I sent you an email. Would be awesome to get this working.
I forgot to add: Using a flip32, all ESCs are detected fine.. so this is specific to the CC3D I think.

**Comment by 4712** [comment]
Now the next step, I hope the final solution:
To test please update to the latest betaflight hex

For the changes I made see here:
https://github.com/4712/cleanflight/commit/d9572353585470894a08f55b8730d476669448c0

and download the BLHeliSuite.exe 1.4.1.0.3.1 and replace it with the existing BLHeliSuite.exe in the BLHeliSuite 1.4.1.0.3 folder. 
http://www.mediafire.com/download/5yd45y0e2st4mry/BLHeliSuite141031.7z

**Comment by 4712** [comment]
For anyone interested: I forked cleanflight v1.10 and merged the BLHeli bootloader support.
Here are the binaries for beta test:
https://github.com/4712/cleanflight/tree/BLHeli-bootloader-support/obj

Download the BLHeliSuite.exe 1.4.1.0.3.1 and replace it with the existing BLHeliSuite.exe in the BLHeliSuite 1.4.1.0.3 folder.
http://www.mediafire.com/download/5yd45y0e2st4mry/BLHeliSuite141031.7z

**Comment by smiling-Jack** [comment]
I test it  today and it is a awesome feature and it works very well. I hope it will be merge in 1.11

**Comment by 4712** [comment]
@hydra, seems to run fine for all, what keeps us from merging this branch?

**Comment by borisbstyle** [comment]
@4712 
It seems to work very nice on naze32, but on my SpracingF3 a bit buggy. Kind of loses connections all the time and multiesc didn't seem to work.
Still would love to see this merged! Very nice feature!

**Comment by 4712** [comment]
@borisbstyle unfortunately I don't have a SpracingF3  to check...

**Comment by nathantsoi** [comment]
fwiw i developed this on the stm32f3discovery board. where do i get a spracingf3? i can debug...

**Comment by CrazyCoder** [comment]
@nathantsoi you can get a cheaper clone at http://www.goodluckbuy.com/racing-f3-flight-controller-deluxe-version-for-fpv-multicopter-qav250.html .

**Comment by 4712** [comment]
Mmmhh...
I remember:
https://github.com/cleanflight/cleanflight/issues/216#issuecomment-103058170

**Comment by hydra** [comment]
@4712 This is the holdup.

https://github.com/cleanflight/cleanflight/pull/1246#issuecomment-138010414

I don't think it will take someone log to fix up the code to remove the target specific stuff as I suggested.

@nathantsoi  "We can pay down the technical debt later, but there is no sense in delaying a good feature while it's re-written"

It doesn't need to be re-written, it's just a cleanup.  I disagree about the tech-debt - who will do it?  burden on me?  Often what happens is code gets added and the never cleaned up.  At the moment there is enough old code to cleanup still without adding to that task with new code that needs even more cleanup.

Can you have a crack at it, give me a shout if my idea is a no-go or there are other unforseen problems.

**Comment by nathantsoi** [comment]
thx for all your hard work keeping the code clean, @hydra! i just think the longer we wait to merge this in, the harder it's going to be, plus it's tested in the current state and working for almost everyone, except the spracing f3 board which should be working soon (thx again @MassiveOverkill!), so there is some risk of breakage refactoring it...

e.g. ppl writing code against this pr: https://github.com/borisbstyle/cleanflight/commit/18aaf7a8924160c770a8f88079cda3cad7d5fe80, if we re-factored i'd have to find all these and re-write as well and forks couldn't merge any updates from the feature

happy to help with the code cleanup as i can. dont have much free time at the moment, so i try to maximize my coding minutes, putting them where they count the most, which is mostly on these [Android fpv goggles](https://plus.google.com/communities/118127535984046831266) at the moment. maybe the next feature/pr i can get your tips (like in https://github.com/cleanflight/cleanflight/pull/1246#issuecomment-138010414) before i start.

keep up the great work! the beers' on me if your ever in [the bay area](https://goo.gl/maps/oQAdyz6DL8A2) :)

**Comment by rotarypower101** [comment]
May I ask what may be a silly question?

Could the same type of process or perhaps something vaguely similar work on a OSD like the microminimOSD for a passthrough?

Does the proper hardware exist to essentially have a serial adaptor through the UART ports so the microminim could stay mounted, and still be able to change settings without disassembly?

To me this type of thinking of tightly integrating components is a monumental change in what could be, I really hope it continues in this direction!

Even to just have the structure or roadmap in place to allow such things would be fantastic!

Thank you for all the hard work you guys put into this hobby!

**Comment by dkisselev** [comment]
@rotarypower101 It's definitely possible, CF already supports GPS passthrough which does the same kind of serial passthrough that an OSD would need, so it shouldn't take much code to add the ability for arbitrary serial passthrough (it might be possible to just rename gps passthrough to serial passthrough and modify the configuration a bit to allow the user to configure the port they want to pass instead of the GPS port as is hardcoded now).

That being said, a lot of IO-based updates are currently being held up by that IO mapping issue, so I'm not sure if anyone would take it on until that's done.

I would suggest opening a new issue to request the feature/improvement so we can track progress and maybe someone will take it on, since it's off-topic here.

**Comment by hydra** [comment]
Merged.  Cleanup will have to happen later.  This is a very popular feature, the lack of which in cleanflight itself is splitting hte user base.

I'll state for the record I'm not happy with the target specific code as per my comment https://github.com/cleanflight/cleanflight/pull/1246#issuecomment-147489183

@nathantsoi Please can you address this with a new PR?

**Comment by simonvc** [comment]
Is this confirmed working on the SPR yet? (i can test tomorrow and confirm)

**Comment by hydra** [comment]
@simonvc not by me.  Likely i will do some cleanup and additional testing on this soon.  We can always revert if there's issues.

**Comment by hydra** [comment]
@rotarypower101 please raise a feature request for your idea of 'generic serial passthough' there is certainly merit in that.

**Comment by 4712** [comment]
@hydra did you manage to send me a board?

**Comment by hydra** [comment]
alas no, please drop me an email with your address again and I'll get one to you tomorrow.

**Comment by 4712** [comment]
Thanks, done.

**Comment by 4712** [comment]
I git the chance to check with RMDO and added a PR https://github.com/cleanflight/cleanflight/pull/1465. 
To work with all 8 ESC (instead of 6), BLHeliSuite rev.14.2.02 will be released asap.
I could NOT test with the SPRacingF3 yet...

**Comment by 4712** [comment]
Thanks a lot @Hydra, I just received the SPRacingF3 and tested it immediately. The code in the PR is running 1wire pass through on the SPRacingF3 without any issues.

**Pull request merged**